hikari/impl/bot.py

```diff
@@ -23,589 +23,481 @@
+        *,
+        activity: typing.Optional[presences.Activity] = None,
+        afk: bool = False,
+        idle_since: typing.Optional[datetime.datetime] = None,
+        ignore_session_start_limit: bool = False,
+        large_threshold: int = 250,
+        shard_ids: typing.Optional[typing.Set[int]] = None,
+        shard_count: typing.Optional[int] = None,
+        status: presences.Status = presences.Status.ONLINE,
+    ) -> None:
+        if shard_ids is not None and shard_count is None:
+            raise TypeError("Must pass shard_count if specifying shard_ids manually")
 
-    @property
-    def uptime(self) -> float:
-        """Return the number of seconds that the bot has been up.
+        # Dispatch the update checker, the sharding requirements checker, and dispatch
+        # the starting event together to save a little time on startup.
+        asyncio.create_task(ux.check_for_updates(), name="check for package updates")
+        requirements_task = asyncio.create_task(self._rest.fetch_gateway_bot(), name="fetch gateway sharding settings")
+        await self.dispatch(lifetime_events.StartingEvent(app=self))
+        requirements = await requirements_task
+
+        if shard_count is None:
+            shard_count = requirements.shard_count
+        if shard_ids is None:
+            shard_ids = set(range(shard_count))
+
+        if requirements.session_start_limit.remaining < len(shard_ids) and not ignore_session_start_limit:
+            _LOGGER.critical(
+                "would have started %s session(s), but you only have %s remaining until %s. Starting more sessions "
+                "than you are allowed to start may result in your token being reset. To skip this message, "
+                "use bot.run(..., ignore_session_start_limit=True) or bot.start(..., ignore_session_start_limit=True)"
+            )
+            raise errors.GatewayError("Attempted to start more sessions than were allowed in the given time-window")
+
+        _LOGGER.info(
+            "planning to start %s session%s... you can start %s session%s before the next window starts at %s",
+            len(shard_ids),
+            "s" if len(shard_ids) != 1 else "",
+            requirements.session_start_limit.remaining,
+            "s" if requirements.session_start_limit.remaining != 1 else "",
+            requirements.session_start_limit.reset_at,
+        )
 
-        This is a monotonic time. To get the physical startup date, see
-        `BotApp.started_at` instead.
+        for window_start in range(0, shard_count, requirements.session_start_limit.max_concurrency):
+            window = [
+                candidate_shard_id
+                for candidate_shard_id in range(
+                    window_start, window_start + requirements.session_start_limit.max_concurrency
+                )
+                if candidate_shard_id in shard_ids
+            ]
+
+            if not window:
+                continue
+            if self._shards:
```

Comment:
is there a reason this check is run for every window_start, would this not wipe the previously started shards for this session if there's more than one window?